drill.cpp

```diff
@@ -383,32 +383,46 @@ bool ExcellonProcessor::millhole(std::ofstream &of, double start_x, double start
     // evenly sized passes such that each pass is at most
     // cutter->stepsize in depth.
-    unsigned int stepcount = 1;
+    int stepcount = 1;
     if (cutter->do_steps) {
-        stepcount = (unsigned int) ceil(abs(cutter->zwork / cutter->stepsize));
+        stepcount = (int) ceil(abs(cutter->zwork / cutter->stepsize));
     }
 
+    double delta_x = stop_x - start_x;
+    double delta_y = stop_y - start_y;
+    double distance = sqrt(delta_x*delta_x + delta_y*delta_y);
     if (cutdiameter * 1.001 >= holediameter) { //In order to avoid a "zero radius arc" error
         // Hole is smaller than cutdiameter so just drill/zig-zag.
         of << "G0 X" << start_x * cfactor << " Y" << start_y * cfactor << '\n';
-        if (slot)
-        {
-            for (unsigned int current_step = 0; true;)
-            {
-                double z = double(current_step+1)/(stepcount) * cutter->zwork;
-                of << "G1 Z" << z * cfactor << '\n';
-                of << "G1 X" << stop_x * cfactor << " Y" << stop_y * cfactor << '\n';
-                current_step++;
-                if (current_step >= stepcount) {
-                    break;
+        if (slot) {
+            // Start one step above Z0 for optimal entry
+            of << "G1 Z" << -1.0/stepcount * cutter->zwork * cfactor
+              << " F" << cutter->vertfeed * cfactor << '\n';
+
+            // Is there enough room for material evacuation?
+            if (distance > 0.3 * cutdiameter) {
+                of  << "F" << cutter->feed * cfactor << '\n';
+            }
+
+            double zhalfstep = cutter->zwork / stepcount / 2;
+            for (int current_step = -1; true; current_step++) {
+                // current_step == stepcount is for the bottom pass, so z needs to stay the same
+                double z = double(std::min(stepcount, current_step+1))/stepcount * cutter->zwork;
+                of << "G1 X" << stop_x * cfactor
+                   << " Y" << stop_y * cfactor;
+                if(stepcount != current_step) {
+                   // Drop superfluous Z from the bottom pass to indicate that this line is not there by accident
+                   of << " Z" << (z - zhalfstep) * cfactor;
                 }
-                z = double(current_step+1)/(stepcount) * cutter->zwork;
-                of << "G1 Z" << z * cfactor << '\n';
-                of << "G1 X" << start_x * cfactor << " Y" << start_y * cfactor << '\n';
-                current_step++;
+                of << '\n';
+                // We don't need a second "zag" on the bottom pass
                 if (current_step >= stepcount) {
                     break;
                 }
+                of << "G1 X" << start_x * cfactor
+                   << " Y" << start_y * cfactor
+                   << " Z" << z * cfactor
+                   << '\n';
             }
         } else {
             of << "G1 Z" << cutter->zwork * cfactor << '\n';
```

Comment:
But what about this line?  If we milldrill a slot and then milldrill all the way through, this G1 will use the most recent feedrate, right?  So it'll be at the horizontal rate instead of the vertical one.  Right?

Just need to add an "F" command there.  It's always vertical.

Reply:
You are right. There must be an explicit feedrate there. I will prepare a pull request.